go/vt/sqlparser/sql.y

```diff
@@ -2097,29 +2106,53 @@ zero_fill_opt:
 
 charset_opt:
   {
-    $$ = ""
+    $$ = ColumnCharset{}
   }
-| charset_or_character_set sql_id
+| charset_or_character_set sql_id binary_opt
   {
-    $$ = string($2.String())
+    $$ = ColumnCharset{Name: string($2.String()), Binary: $3}
   }
-| charset_or_character_set STRING
+| charset_or_character_set STRING binary_opt
   {
-    $$ = encodeSQLString($2)
+    $$ = ColumnCharset{Name: encodeSQLString($2), Binary: $3}
   }
 | charset_or_character_set BINARY
   {
-    $$ = string($2)
+    $$ = ColumnCharset{Name: string($2)}
   }
-| ASCII
+| ASCII binary_opt
   {
     // ASCII: Shorthand for CHARACTER SET latin1.
-    $$ = "latin1"
+    $$ = ColumnCharset{Name: "latin1", Binary: $2}
   }
-| UNICODE
+| UNICODE binary_opt
   {
     // UNICODE: Shorthand for CHARACTER SET ucs2.
-    $$ = "ucs2"
+    $$ = ColumnCharset{Name: "ucs2", Binary: $2}
+  }
+| BINARY
```

Comment:
`binary` on its own is also allowed, which means the default character set but with the binary collation.